armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+            writer.close()
+
+        self.closed = True
+
+
+class Meter:
+    def __init__(
+        self,
+        name,
+        metric,
+        *metric_arg_names,
+        metric_kwargs=None,
+        auto_measure=True,
+        final=None,
+        final_name=None,
+        final_kwargs=None,
+        keep_results=True,
+    ):
+        """
+        StandardMeter(metrics.l2, "model.x_post[benign]", "model.x_post[adversarial]")
+
+        metric_kwargs - kwargs that are constant across measurements
+        auto_measure - whether to measure when all of the variables are present
+            if False, 'measure()' must be called externally
+
+        final - metric that takes in list of results as input
+            Example: np.mean
+        final_name - if final is not None, this is the name associated with the record
+            if not specified, it defaults to f'{final}_{name}'
+
+        keep_results - whether to locally store results
+            if final is not None, keep_results is set to True
+        """
+        self.name = str(name)
+        if not callable(metric):
+            raise ValueError(f"metric {metric} must be callable")
+        self.metric = metric
+        if not len(metric_arg_names):
+            log.warning("metric_arg_names is an empty list")
+        self.arg_index = {arg: i for i, arg in enumerate(metric_arg_names)}
+        self.metric_kwargs = metric_kwargs or {}
+        if not isinstance(self.metric_kwargs, dict):
+            raise ValueError(
+                f"metric_kwargs must be None or a dict, not {metric_kwargs}"
+            )
+        self.clear()
+        self._results = []
+        self.writers = []
+        self.auto_measure = bool(auto_measure)
+        self._warned = False
+
+        keep_results = bool(keep_results)
+        if final is None:
+            if final_name is not None:
+                final_name = str(final_name)
+        else:
+            if not callable(final):
+                raise ValueError(f"final {final} must be callable")
+            if final_name is None:
+                final_name = f"{final}_{name}"
+            else:
+                final_name = str(final_name)
+            if not keep_results:
+                log.warning("final is not None. Overriding keep_results to True")
+                keep_results = True
+        self.final = final
+        self.final_name = final_name
+        self.final_kwargs = final_kwargs or {}
+        self.final_result = None
+        if not isinstance(self.final_kwargs, dict):
+            raise ValueError(f"final_kwargs must be None or a dict, not {final_kwargs}")
+        self.keep_results = keep_results
+
+    def get_arg_names(self):
+        return list(self.arg_index)
+
+    def clear(self):
+        """
+        Clear all of the current values
+        """
+        self.values = [None] * len(self.arg_index)
+        self.values_set = [False] * len(self.arg_index)
+        self.batches = [None] * len(self.arg_index)
```

Comment:
the `self.batches` attribute caused me some confusion. If I understand, it's a list of length equal to the number of args. And each element is the batch_index of where the arg came from. So all elements should be the same (i.e. come from same batch). If that's all correct, maybe something like `self.arg_batch_indices` would be more descriptive

Reply:
Yes, that would be more informative.